freqtrade/optimize/recursive_analysis.py

```diff
@@ -0,0 +1,182 @@
+
+        self.partial_varHolder_array: List[VarHolder] = []
+        self.partial_varHolder_lookahead_array: List[VarHolder] = []
+
+        self.dict_recursive: Dict[str, Any] = dict()
+
+    # For recursive bias check
+    # analyzes two data frames with processed indicators and shows differences between them.
+    def analyze_indicators(self):
+
+        pair_to_check = self.local_config['pairs'][0]
+        logger.info("Start checking for recursive bias")
+
+        # check and report signals
+        base_last_row = self.full_varHolder.indicators[pair_to_check].iloc[-1]
+
+        for part in self.partial_varHolder_array:
+            part_last_row = part.indicators[pair_to_check].iloc[-1]
+
+            compare_df = base_last_row.compare(part_last_row)
+            if compare_df.shape[0] > 0:
+                # print(compare_df)
+                for col_name, values in compare_df.items():
+                    # print(col_name)
+                    if 'other' == col_name:
+                        continue
+                    indicators = values.index
+
+                    for indicator in indicators:
+                        if (indicator not in self.dict_recursive):
+                            self.dict_recursive[indicator] = {}
+
+                        values_diff = compare_df.loc[indicator]
+                        values_diff_self = values_diff.loc['self']
+                        values_diff_other = values_diff.loc['other']
+                        diff = (values_diff_other - values_diff_self) / values_diff_self * 100
+
+                        self.dict_recursive[indicator][part.startup_candle] = f"{diff:.3f}%"
+
+            else:
+                logger.info("No difference found. Stop the process.")
+                break
+
+    # For lookahead bias check
+    # analyzes two data frames with processed indicators and shows differences between them.
+    def analyze_indicators_lookahead(self):
+
+        pair_to_check = self.local_config['pairs'][0]
+        logger.info("Start checking for lookahead bias on indicators only")
+
+        part = self.partial_varHolder_lookahead_array[0]
+        part_last_row = part.indicators[pair_to_check].iloc[-1]
+        date_to_check = part_last_row['date']
+        index_to_get = (self.full_varHolder.indicators[pair_to_check]['date'] == date_to_check)
+        base_row_check = self.full_varHolder.indicators[pair_to_check].loc[index_to_get].iloc[-1]
+
+        check_time = part.to_dt.strftime('%Y-%m-%dT%H:%M:%S')
+
+        logger.info(f"Check indicators at {check_time}")
+        # logger.info(f"vs {part_timerange} with {part.startup_candle} startup candle")
+
+        compare_df = base_row_check.compare(part_last_row)
+        if compare_df.shape[0] > 0:
+            # print(compare_df)
+            for col_name, values in compare_df.items():
+                # print(col_name)
+                if 'other' == col_name:
+                    continue
```

Comment:
what's the reason to exclude `'other'` from this test ?

Reply:
we skip 'other' because `.compare` will create 2 columns  'self' and 'others', and we only need to go through it just once

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.compare.html

Reply:
ah ok perfect - thanks for the explanation